docs/src/main/asciidoc/hibernate-orm.adoc

```diff
@@ -1367,6 +1370,60 @@ and annotating it with `@PersistenceUnitExtension` (or `@PersistenceUnitExtensio
 will replace the current Quarkus default implementation `io.quarkus.hibernate.orm.runtime.tenant.DataSourceTenantConnectionResolver`.
 Your custom connection resolver would allow for example to read tenant information from a database and create a connection per tenant at runtime based on it.
 
+[CAUTION]
+====
+<<automatic-integration,Automatic integrations>> will **not** work with programmatically-created `ConnectionProvider`.
+This implies all these integrations, including those from other Quarkus modules (e.g. xref:smallrye-health.adoc[SmallRye Health]) must be done manually.
+====
+
+Here is an implementation example of a `TenantConnectionResolver` implementation using standard Quarkus technologies – Agroal for pooling and Narayana for transactions:
+
+[source,java]
+----
+@ApplicationScoped
+@PersistenceUnitExtension
+public class ExampleTenantConnexionResolver implements TenantConnectionResolver {
+
+    private final jakarta.transaction.TransactionManager transactionManager;
+    private final TransactionSynchronizationRegistry transactionSynchronizationRegistry;
+
+    public ExampleTenantConnexionResolver(
+            TransactionManager transactionManager,
+            TransactionSynchronizationRegistry transactionSynchronizationRegistry) {
+        this.transactionManager = transactionManager;
+        this.transactionSynchronizationRegistry = transactionSynchronizationRegistry;
+    }
+
+    @Override
+    public ConnectionProvider resolve(String tenantId) {
+        return new QuarkusConnectionProvider(createDatasource(tenantId));
```

Comment:
`QuarkusConnectionProvider` is technically an internal (non-API) class... We shouldn't recommend using it in the documentation.

Reply:
OK, but what would be the correct way to provide a `ConnectionProvider` here?

Here are all implementations known of this interface in my project, all of them seem private/internal. Should everyone that use custom datasources implement their own `ConnectionProvider`?

![](https://github.com/user-attachments/assets/03961cfa-ae36-4140-a994-9774fbb984c0)

Reply:
> Here are all implementations known of this interface in my project, all of them seem private/internal. Should everyone that use custom datasources implement their own `ConnectionProvider`?

Lacking implementations in the API (!), I think so, yes. Fortunately that's not a _lot_ of code.

I mean, `QuarkusConnectionProvider` will most likely work fine for the foreseeable future... But since this is documentation, it's the main source of truth, so I think we shouldn't recommend anything even slightly dodgy?

The alternative is that we add a `ConnectionProvider` implementation in the API, but that seems a bit out of scope.

Slightly off-topic note: https://github.com/quarkusio/quarkus/issues/11861 would make the whole thing _much_ easier to use and to document...

Reply:
> I can just replace the `QuarkusConnectionProvider` by a `YourOwnCustomConnectionProviderImpl` in the documentation if this is OK for you

Thanks!

Reply:
Done